### PR TITLE
[14.0] [FIX] product_state

### PR DESCRIPTION
**Pull request by sonhd91**
* Fix bug when inverting non-exist new state
* Show default and active fields in product state form
* Refactor inverse state to use with variant state

### CODE REVIEW

File: product_state/models/product_template.py [hunk 1]
```diff
@@ -35,22 +35,26 @@ def _inverse_product_state_id(self):
 
     @api.model
     def _get_default_product_state_id(self):
-        return self.env["product.state"].search([("default", "=", True)], limit=1).id
+        return self.env["product.state"].search([("default", "=", True)], limit=1)
```
Comment (simahawk) [inline product_state/models/product_template.py:39]: bad naming of the method then :)

File: product_state/models/product_template.py [hunk 2]
```diff
@@ -35,22 +35,26 @@ def _inverse_product_state_id(self):
...
             product_tmpl.state = product_tmpl.product_state_id.code
 
     def _inverse_product_state(self):
-        ProductState = self.env["product.state"]
         for product_tmpl in self:
-            product_state = ProductState.search(
-                [("code", "=", product_tmpl.state)], limit=1
+            self._set_product_state_id(product_tmpl)
+
+    def _set_product_state_id(self, record):
```
Comment (simahawk) [inline product_state/models/product_template.py]:
```suggestion
            product_tmpl._set_product_state_id()

    def _set_product_state_id(self):
```

File: product_state/models/product_template.py [hunk 3]
```diff
@@ -35,22 +35,26 @@ def _inverse_product_state_id(self):
...
+        """ The record param is for similar state field at product.product model. """
+        ProductState = record.env["product.state"]
+        product_state = ProductState.search([("code", "=", record.state)], limit=1)
+        if record.state and not product_state:
+            product_state = ProductState.create(
```
Comment (simahawk) [inline product_state/models/product_template.py]: I wouldn't do this :thinking:

**Review by simahawk — CHANGES_REQUESTED** [review]
I don't get what's the point of creating a state on the fly

File: product_state/views/product_template.xml [hunk 4]
```diff
@@ -57,6 +57,8 @@
                     </group>
                     <group string="Description">
                         <field name="description" nolabel="1" />
+                        <field name="default" widget="boolean_toggle" />
+                        <field name="active" widget="boolean_toggle" />
                     </group>
```
Comment (i-vyshnevska) [inline product_state/views/product_template.xml:62]: pls add a fix on line 26
` <field name="domain">[('id', '=', product_state_id)]</field>`

File: product_state/models/product_template.py (hunk 1 above)
Comment (sonhd91) [reply product_state/models/product_template.py:39]: As we are removing the method from `product_tier_validation` (https://github.com/OCA/product-attribute/pull/962), we can rename it.

File: product_state/models/product_template.py (hunk 2 above)
Comment (sonhd91) [reply product_state/models/product_template.py]: The method is being used by both `product.template` and `product.product` so `record` is needed. It is now promoted to `api.model` level.

File: product_state/models/product_template.py (hunk 3 above)
Comment (sonhd91) [reply product_state/models/product_template.py]: Thanks for the advice, it has been fixed.

File: product_state/views/product_template.xml (hunk 4 above)
Comment (sonhd91) [reply product_state/views/product_template.xml:62]: Thanks for the advice, it has been fixed.

**Comment by hailangvn** [comment]
Thanks @simahawk  and @i-vyshnevska for your review. The PR has been updated.

File: product_state/models/product_template.py [hunk 5]
```diff
@@ -22,7 +23,7 @@ class ProductTemplate(models.Model):
         help="Select a state for this product",
         group_expand="_read_group_state_id",
         inverse="_inverse_product_state_id",
-        default=lambda self: self._get_default_product_state_id(),
+        default=lambda self: self._get_default_product_state(),
```
Comment (i-vyshnevska) [inline product_state/models/product_template.py]:
```suggestion
        default=lambda self: self._get_default_product_state().id,
```

**Review by i-vyshnevska — APPROVED** [review]

**Review by simahawk — APPROVED** [review]

**Pull request merged**